pkg/gadgets/bindsnoop/gadget.go

```diff
@@ -0,0 +1,189 @@
+	}
+}
+
+func (f *TraceFactory) Description() string {
+	return `bindsnoop traces the kernel functions performing socket binding.`
+}
+
+func (f *TraceFactory) OutputModesSupported() map[string]struct{} {
+	return map[string]struct{}{
+		"Stream": {},
+	}
+}
+
+func deleteTrace(name string, t interface{}) {
+	trace := t.(*Trace)
+	if trace.tracer != nil {
+		trace.tracer.Stop()
+	}
+}
+
+func (f *TraceFactory) Operations() map[string]gadgets.TraceOperation {
+	n := func() interface{} {
+		return &Trace{
+			resolver: f.Resolver,
+		}
+	}
+
+	return map[string]gadgets.TraceOperation{
+		"start": {
+			Doc: "Start bindsnoop gadget",
+			Operation: func(name string, trace *gadgetv1alpha1.Trace) {
+				f.LookupOrCreate(name, n).(*Trace).Start(trace)
+			},
+		},
+		"stop": {
+			Doc: "Stop bindsnoop gadget",
+			Operation: func(name string, trace *gadgetv1alpha1.Trace) {
+				f.LookupOrCreate(name, n).(*Trace).Stop(trace)
+			},
+		},
+	}
+}
+
+func (t *Trace) Start(trace *gadgetv1alpha1.Trace) {
+	if t.started {
+		trace.Status.State = "Started"
+		return
+	}
+
+	traceName := gadgets.TraceName(trace.ObjectMeta.Namespace, trace.ObjectMeta.Name)
+
+	eventCallback := func(event types.Event) {
+		r, err := json.Marshal(event)
+		if err != nil {
+			log.Warnf("Gadget %s: error marshalling event: %s", trace.Spec.Gadget, err)
+			return
+		}
+		t.resolver.PublishEvent(traceName, string(r))
+	}
+
+	params := trace.Spec.Parameters
+
+	targetPid := int32(0)
+	if pid, ok := params["pid"]; ok {
+		pidParsed, err := strconv.ParseInt(pid, 10, 32)
+		if err != nil {
+			trace.Status.OperationError = fmt.Sprintf("%q is not valid for PID", pid)
+			return
+		}
+
+		targetPid = int32(pidParsed)
+	}
+
+	targetPorts := make([]uint16, 0)
+	if portsString, ok := params["ports"]; ok && len(portsString) > 0 {
+		portsStringSlice := strings.Split(portsString, ",")
+
+		for _, portString := range portsStringSlice {
+			portParsed, err := strconv.ParseUint(portString, 10, 16)
+			if err != nil {
+				trace.Status.OperationError = fmt.Sprintf("%q is not valid for port", portString)
+				return
+			}
+
+			targetPorts = append(targetPorts, uint16(portParsed))
+		}
+	}
+
+	ignoreErrors := false
+	if ignore, ok := params["ignore_errors"]; ok {
+		ignoreErrorsParsed, err := strconv.ParseBool(ignore)
+		if err != nil {
+			trace.Status.OperationError = fmt.Sprintf("%q is not valid for ignore", ignore)
+			return
+		}
+
+		ignoreErrors = ignoreErrorsParsed
+	}
+
+	var err error
+
+	config := &tracer.Config{
+		MountnsMap:   gadgets.TracePinPath(trace.ObjectMeta.Namespace, trace.ObjectMeta.Name),
+		TargetPid:    targetPid,
+		TargetPorts:  targetPorts,
+		IgnoreErrors: ignoreErrors,
+	}
+	t.tracer, err = coretracer.NewTracer(config, t.resolver, eventCallback, trace.Spec.Node)
+	if err != nil {
+		// TODO: The following line causes the client to fail.
+		// trace.Status.OperationWarning = fmt.Sprint("failed to create core tracer. Falling back to standard one")
```

Comment:
This is now fixed by https://github.com/kinvolk/inspektor-gadget/pull/527, so you can uncomment that line.